### PR TITLE
wtclient: queue backup id only

### PR DESCRIPTION
**Pull request by ellemouton**
Currently, the entire breach retribution struct is passed to the watchtower client 
and queued in its task pipeline. This is no longer necessary with the changes made 
in https://github.com/lightningnetwork/lnd/pull/7379 since it is now possible for the
tower client to build the breach retribution info on the fly.

So what this PR does is to now only queue the `wtdb.BackupID` of the update in the
task pipeline and reconstruct the breach retribution only at the time that it will be 
required. 

This PR sets up all the ground work that will be required for https://github.com/lightningnetwork/lnd/pull/7380

## Flow of the PR

1. A new `FetchChannelByID` method is added to `ChannelStateDB` to make it
possible to later on fetch a channel by its `lnwire.ChannelID`
representation. The reason for this is that currently there is only a
`FetchChannel` method which allows you to fetch a channel by its `wire.Outpoint`
but the tower client uses `lnwire.ChannelID` to keep track of channels so we
need a new lookup method here.
2. To prepare for the later on when we want to be able to construct the
BreachRetribution at the time that it is needed, we create and pass through a
`buildBreachRetribution` callback to the tower client.
3. Since the tower client now has the ability to build the breach retribution by
itself, we let it do that instead of having the `channellink` do that. This
is to prepare for later on when we construct the breach info JIT.
4. Move the construction of the breach retribution to the time of binding it to
a session.
5. At this point, we can now let the (still in-memory) pipeline task queue only
carry the `BackupID`.

### CODE REVIEW

**Comment by ellemouton** [comment]
Hey @guggero  & @bitromortac :) 

This is a split off from https://github.com/lightningnetwork/lnd/pull/7380 so I think mostly reviewed already :)

**Comment by ellemouton** [comment]
Thanks for the review @bitromortac 🙏 

> Do you have an estimate how much compression we get by this per backup

The `wtdb.BackupID` is a fixed 40 bytes while the `lnwallet.BreachRetribution` is muuuch bigger: a quick calculation estimates that the lower bound of bytes it holds is: 510 + 94(numHTLCs). And this is without taking into account any `[]byte` fields.

File: channeldb/db.go [hunk 1]
```diff
@@ -658,19 +659,42 @@ func (c *ChannelStateDB) fetchNodeChannels(chainBucket kvdb.RBucket) (
 func (c *ChannelStateDB) FetchChannel(tx kvdb.RTx, chanPoint wire.OutPoint) (
 	*OpenChannel, error) {
 
+	selector := func(chainBkt walletdb.ReadBucket) ([]byte, *wire.OutPoint,
+		error) {
+
+		var targetChanPoint bytes.Buffer
+		err := writeOutpoint(&targetChanPoint, &chanPoint)
```
Comment (guggero) [inline channeldb/db.go:666]: This should be done outside of the callback to save a couple of bytes in allocation.

File: channeldb/db.go [hunk 2]
```diff
@@ -674,6 +674,54 @@ func (c *ChannelStateDB) FetchChannel(tx kvdb.RTx, chanPoint wire.OutPoint) (
 	return c.channelScanner(tx, selector)
 }
 
+// FetchChannelByID attempts to locate a channel specified by the passed channel
+// ID. If the channel cannot be found, then an error will be returned.
+// Optionally an existing db tx can be supplied.
+func (c *ChannelStateDB) FetchChannelByID(tx kvdb.RTx, id lnwire.ChannelID) (
+	*OpenChannel, error) {
+
+	selector := func(chainBkt walletdb.ReadBucket) (
```
Comment (guggero) [inline channeldb/db.go:683]: nit: format the same as the one in the previous commit?

File: channeldb/db.go [hunk 3]
```diff
@@ -674,6 +674,54 @@ func (c *ChannelStateDB) FetchChannel(tx kvdb.RTx, chanPoint wire.OutPoint) (
...
+		)
+		err := chainBkt.ForEach(func(k, _ []byte) error {
+			var outPoint wire.OutPoint
+			err := readOutpoint(bytes.NewReader(k), &outPoint)
+			if err != nil {
+				return err
+			}
+
+			chanID := lnwire.NewChanIDFromOutPoint(&outPoint)
+			if !bytes.Equal(chanID[:], id[:]) {
```
Comment (guggero) [inline channeldb/db.go:703]: Both variables are of type `[32]byte`, which are directly comparable by `==`.

File: channeldb/db_test.go [hunk 4]
```diff
@@ -252,9 +257,11 @@ func TestFetchChannel(t *testing.T) {
...
 
 	_, err = cdb.FetchChannel(nil, channelState2.FundingOutpoint)
-	if err == nil {
-		t.Fatalf("expected query to fail")
-	}
+	require.Error(t, err)
+
+	chanID2 := lnwire.NewChanIDFromOutPoint(&channelState2.FundingOutpoint)
+	_, err = cdb.FetchChannelByID(nil, chanID2)
+	require.Error(t, err)
```
Comment (guggero) [inline channeldb/db_test.go:264]: Assert the error type?

**Review by guggero — COMMENTED** [review]
Thank you for extracting those commits out. Almost ready, just a few small nits.

File: channeldb/db.go (hunk 1 above)
Comment (ellemouton) [reply channeldb/db.go:666]: noice 🔥

**Pull request merged**